ES8: Add migration logic for mirroring

rebeccahum opened the pull request:
## Description

This PR implements double-write functionality for Elasticsearch migration, ensuring data consistency by simultaneously writing to both Elasticsearch 7 and Elasticsearch 8 clusters during the migration process.

When migrating from ES7 to ES8, this implementation ensures zero data loss by writing all index operations to both the existing ES7 cluster and the new ES8 cluster simultaneously.

### How Double-Write Works

1. Migration Detection: System detects migration state via `VIP_ELASTICSEARCH_MIGRATION_IN_PROGRESS` constant
2. Dual Endpoint Writing: All write operations (index, bulk_index, delete) are sent to:
    - Primary ES7 endpoints (existing production data)
    - Migration ES8 endpoints (new cluster being populated (`VIP_ELASTICSEARCH_MIGRATION_ENDPOINTS`))
3. Failure Handling: If ES8 write fails, ES7 continues normally to maintain service availability



## Pre-review checklist

Please make sure the items below have been covered before requesting a review:

- [ ] This change works and has been tested locally or in Codespaces (or has an appropriate fallback).
- [ ] This change works and has been tested on a sandbox.
- [ ] This change has relevant unit tests (if applicable).
- [ ] This change uses a rollout method to ease with deployment (if applicable - especially for large scale actions that require writes).
- [ ] This change has relevant documentation additions / updates (if applicable).
- [ ] I've created a changelog description that aligns with the provided examples.

## Pre-deploy checklist

- [ ] VIP staff: Ensure any alerts added/updated conform to internal standards (see internal documentation). 

## Steps to Test

<!--
Outline the steps to test and verify the PR here.

Example:

1. Check out PR.
1. Go to `wp-admin` > `Tools` > `Bakery`
1. Click on "Bake Cookies" button.
1. Verify cookies are delicious.
-->

rebeccahum committed:
ES8: Add migration logic for mirroring

rinatkhaziev commented on search/includes/classes/class-search.php:
```diff
@@ -1037,6 +1040,67 @@ public function filter__ep_do_intercept_request( $request, $query, $args, $failu
+			return;
+		}
+
+		foreach ( $migration_hosts as $migration_host ) {
+			$migration_url = preg_replace(
```
I think this needs a random pick vs iterating over each host, cause that's a request per each one.

rebeccahum committed:
Add check if in migration mode

luismulinari commented on search/includes/classes/class-search.php:
```diff
@@ -1037,6 +1040,66 @@ public function filter__ep_do_intercept_request( $request, $query, $args, $failu
+	protected function mirror_write_to_migration_hosts( $query, $args, $type = null ) {
+		if ( ! defined( 'VIP_ELASTICSEARCH_VERSION' ) || constant( 'VIP_ELASTICSEARCH_VERSION' ) !== '8' ) {
+			return;
+		}
```
```suggestion
```

I think we should remove this constant and rely only on `VIP_ELASTICSEARCH_MIGRATION_ENDPOINTS` and `VIP_ELASTICSEARCH_MIGRATION` to decide if we need to deploy-write.

luismulinari committed:
refactor(search): improve ES8 migration constant naming and logic

- Rename VIP_ELASTICSEARCH_MIGRATION to VIP_ELASTICSEARCH_MIGRATION_IN_PROGRESS for clarity
- Simplify migration mirroring logic by consolidating condition checks
- Add comprehensive tests for migration double-write functionality
- Fix test helper functions to properly handle function arguments

luismulinari committed:
Merge branch 'develop' into es/8-migration-mirroring

luismulinari committed:
Fix elasticsearch version condition to define migration endpoints
